src/cluster_import.c

```diff
@@ -0,0 +1,1927 @@
+        return;
+    case SLOT_MIGRATION_EXPORT:
+        clusterHandleSlotExportLinkClientClose(link);
+        return;
+    }
+}
+
+void initClusterSlotMigrationLinkList(void) {
+    server.cluster->slot_migration_links = listCreate();
+    listSetFreeMethod(server.cluster->slot_migration_links, freeSlotMigrationLink);
+}
+
+int shouldCleanupSlotMigrationLink(slotMigrationLink *link, unsigned long idx) {
+    return (link->state == SLOT_MIGRATION_LINK_CANCELLED ||
+            link->state == SLOT_MIGRATION_LINK_FAILED ||
+            link->state == SLOT_MIGRATION_LINK_SUCCESS) &&
+           ((server.cluster_slot_migration_log_ttl &&
+             server.unixtime - link->last_update >= server.cluster_slot_migration_log_ttl) ||
+            idx >= server.cluster_slot_migration_log_max_len);
+}
+
+const char *slotMigrationLinkStateToString(slotMigrationLinkState state) {
+    switch (state) {
+    case SLOT_IMPORT_RECONNECT: return "reconnecting";
+    case SLOT_IMPORT_CONNECTING: return "connecting";
+    case SLOT_IMPORT_AUTHENTICATING: return "authenticating";
+    case SLOT_IMPORT_START_SNAPSHOT:
+    case SLOT_IMPORT_RECEIVE_SNAPSHOT: return "snapshotting";
+    case SLOT_IMPORT_RECEIVE_STREAM: return "replicating";
+    case SLOT_IMPORT_FAILOVER_WAITING_FOR_PAUSED: return "failover-syncing";
+    case SLOT_IMPORT_FAILOVER_REQUESTED: return "failover-requested";
+    case SLOT_IMPORT_FAILOVER_GRANTED: return "failover-granted";
+    case SLOT_IMPORT_FINISHED_WAITING_TO_CLEANUP: return "cleaning-up";
+
+    case SLOT_EXPORT_WAITING_TO_SNAPSHOT: return "waiting-to-snapshot";
+    case SLOT_EXPORT_SNAPSHOTTING: return "snapshotting";
+    case SLOT_EXPORT_STREAMING: return "replicating";
+    case SLOT_EXPORT_FAILOVER_PAUSED: return "failover-paused";
+    case SLOT_EXPORT_FAILOVER_GRANTED: return "failover-granted";
+
+    case SLOT_MIGRATION_LINK_SUCCESS: return "success";
+    case SLOT_MIGRATION_LINK_CANCELLED: return "cancelled";
+    case SLOT_MIGRATION_LINK_FAILED: return "failed";
+    }
+    return "unknown";
+}
+
+void updateSlotMigrationLinkState(slotMigrationLink *link, slotMigrationLinkState state) {
+    serverLog(LL_DEBUG,
+              "Slot %s link %.40s state transition: %s -> %s",
+              link->type == SLOT_MIGRATION_IMPORT ? "IMPORT" : "EXPORT",
+              link->linkname,
+              slotMigrationLinkStateToString(link->state),
+              slotMigrationLinkStateToString(state));
+    link->last_update = server.unixtime;
+    link->state = state;
+}
+
+void finishSlotMigrationLink(slotMigrationLink *link, slotMigrationLinkState state, char *message) {
+    serverLog(LL_NOTICE,
+              "Slot %s link %.40s (for slots [%s]) finished. State: %s, Message: %s",
+              link->type == SLOT_MIGRATION_IMPORT ? "import" : "export",
+              link->linkname,
+              link->slot_ranges_str,
+              slotMigrationLinkStateToString(state), message ? message : "");
```

Comment:
Personally, I would be annoyed by a log entry that ends with "Message: " and no message. Can we either put something like "none" as the message or remove the "Message:" heading when there's no message?

```suggestion
              slotMigrationLinkStateToString(state), message ? message : "none");
```

Reply:
Done